letsencrypt/tests/auth_handler_test.py

```diff
@@ -148,6 +152,66 @@ def test_perform_failure(self):
         self.assertRaises(
             errors.AuthorizationError, self.handler.get_authorizations, ["0"])
 
+    @mock.patch("acme.challenges.HTTP01Response.simple_verify")
+    @mock.patch("acme.challenges.TLSSNI01Response.simple_verify")
+    @mock.patch("letsencrypt.auth_handler.logger")
+    def test_simple_verify(self, mock_logger, mock_tlssni01_verify, mock_http01_verify):
+        account_key = mock.Mock()
+        challenge_http_01 = achallenges.KeyAuthorizationAnnotatedChallenge(
+            challb=acme_util.HTTP01_P, domain=b'localhost', account_key=account_key)
+        challenge_tls_sni_01 = achallenges.KeyAuthorizationAnnotatedChallenge(
+            challb=acme_util.TLSSNI01_P, domain=b'localhost', account_key=account_key)
+
+        response_http_01 = challenges.HTTP01Response(key_authorization=u'foo')
+        response_tls_sni_01 = challenges.TLSSNI01Response(key_authorization=u'foo')
+
+        # Test _simple_verify's behaviour when implementation of challenge is not implemented
+        response_dummy = mock.Mock()
+        challenge_dummy = mock.Mock(chall=mock.Mock(), uri="foo", typ="Mock")
+
+        self.handler.dv_c = [challenge_http_01,
+                             challenge_tls_sni_01,
+                             challenge_dummy]
+
+        self.mock_dv_auth.perform.side_effect = [[response_http_01,
+                                                  response_tls_sni_01,
+                                                  response_dummy]]
+
+        mock_http01_verify.return_value = False
+        mock_tlssni01_verify.return_value = False
```

Comment:
My bogus implementation above might also clarify this a bit hopefully. Success or failure of the verify call is the most important part of this code so it makes sense to test both situations.

You mention that if someone adds a new `logger.warning` it might break if you test the successful verification. This is true and should be avoided by checking for absence of the warning about verification only.

You also mention that the functionality of `simple_verify` is tested but I assume you mean `_simple_verify`? That is (partly) what should be tested. But your changes to `_solve_challenges` should also be tested and right now you are not really testing the return value of `_simple_verify`.

Lets look at the code under test again:
```python
                    for achall, response in zip(self.dv_c, dv_resp):
                        if not self._simple_verify(achall, response):
                            logger.warning(
                                "Self-verify of %s challenge (%s) "
                                "for domain %s failed.",
                                achall.chall.__class__.__name__,
                                achall.domain, achall.uri)
```
This is equivalent to:
```python
                    for achall, response in zip(self.dv_c, dv_resp):
                        retval = self._simple_verify(achall, response)
                        if not retval:
                            logger.warning(
                                "Self-verify of %s challenge (%s) "
                                "for domain %s failed.",
                                achall.chall.__class__.__name__,
                                achall.domain, achall.uri)
```

In this last bit, the line `if not retval:` is at the core of how the code is supposed to work, but it is not properly tested. It can be replaced with many other lines that are wrong but pass this test, for example `if achall.domain == 'localhost':` or `if hasattr(response, 'simple_verify'):` etc.